src/software/ai/evaluation/q_learning/linear_q_function.hpp

```diff
@@ -0,0 +1,244 @@
+#pragma once
+
+#include <Eigen/Dense>
+
+#include "csv.hpp"
+#include "proto/q_learning.pb.h"
+#include "software/ai/evaluation/q_learning/feature_extractor.hpp"
+#include "software/ai/evaluation/q_learning/q_function.hpp"
+
+/**
+ * Implementation of a Q-function with linear function approximation.
+ *
+ * We approximate the Q-function using a linear combination of state features and
+ * their weights. This provides a reasonable estimate of Q(S, A) even if we have not
+ * applied action A in state S previously.
+ *
+ * See https://gibberblot.github.io/rl-notes/single-agent/function-approximation.html
+ * for more details.
+ *
+ * @tparam TState the type representing the state of the MDP
+ * @tparam TAction the type representing the set of actions the agent can execute
+ */
+template <typename TState, typename TAction>
+class LinearQFunction : public QFunction<TState, TAction>
+{
+   public:
+    /**
+     * Creates a LinearQFunction.
+     *
+     * @param features the feature extractor to use on the state representation
+     * @param learning_rate the initial learning rate
+     * @param discount_factor the initial discount factor
+     * @param weights initial weights to use, or std::nullopt to initialize weights with 0
+     */
+    explicit LinearQFunction(FeatureExtractor<TState, TAction> features,
+                             double learning_rate, double discount_factor,
+                             std::optional<Eigen::VectorXd> weights = std::nullopt);
+
+    /**
+     * Creates a LinearQFunction with its weights initialized using the given
+     * CSV file.
+     *
+     * @param features the feature extractor to use on the state representation
+     * @param learning_rate the initial learning rate
+     * @param discount_factor the initial discount factor
+     * @param weights_csv_file the CSV file containing the initial weights to use
+     */
+    explicit LinearQFunction(FeatureExtractor<TState, TAction> features,
```

Comment:
nit: most of the `std::string` args in this file could be `const std::string&` but I doubt that the copy results in a big runtime cost